### PR TITLE
CMake build definitions

### LINKED ISSUE #1635: Migrate pkg-config support to CMake
See pocoproject/poco#2224 for example of using cmake for pkg-config files, support originally added in #860

### PR DESCRIPTION
**Pull request by jackgerrits**
This PR add CMake build definitions. They do not include clr or csharp, but they do build on Windows and Linux. This attempt has gone down the route of as small a change as possible but still a step in the right direction.

Items still left to be done:
- [x] Create base makefile to perform CMake actions for discovery
- [x] Remove Automake
- [x] Remove extra work done to output both shared and static libs at the same time and just use normal CMake behavior
- [x] Update Readme with new workflows and to remove automake references
- [x] Test with and without vcpkg on Linux
- [x] Test and document Clang support 
- [x] Test CMake options for Debug etc
- [x] Include Java and Python installs into CMake install
- [x] Add Num processors to `cmake --build` -- __Not feasible__
- [x] Change tests from custom_targets to proper CMake tests
- [x] Limit build configuration types to Debug and Release
- [x] Update versioning to work with CMake and Windows builds
- [x] Convert Java to CMake
- [x] Update Travis build to use CMake
More:
- [x] Build doesn't seem to understand the number of processors.
- [x] Add targets for profiling, valgrind, static.
- [x] remove ffast-math from valgrind profile (fixing an inconsistency)
- [x] Fix static compile
- [x] Make it so install headers are automatically headers
- [x] Do we need disambiguators for else() in the vowpalwabbit directory?

Fixes #1635

### CODE REVIEW

**Comment by jackgerrits** [comment]
TODO - remove automake

**Comment by JohnLangford** [comment]
This pull request **introduces 9 alerts** when merging c0d19cbec1202d31acfabd227e8eaefb4f1b3941 into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-5824628c8bc5236ed9382253dbe7f33f3f65055d)

**new alerts:**

* 6 for Unused import
* 1 for Unused local variable
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 9 alerts** when merging 437a1e1b869fcc18ad6060f4017c577f0a7ac02c into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-da9eeb7bb8e652ce79df805ddbe47a40c20c5600)

**new alerts:**

* 6 for Unused import
* 1 for Unused local variable
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 9 alerts** when merging b96642539245c48676c290df2b22a89a4963a991 into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-7bc44aeedc371755391a0aa76e19ddca62eb1f2f)

**new alerts:**

* 6 for Unused import
* 1 for Unused local variable
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 14 alerts** when merging ba8517f971fce4e27cf50964bb1284dcbed4d8b0 into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-69cf34600164c95d927ed6bb687bf7d6a4554133)

**new alerts:**

* 6 for Unused import
* 2 for Local variable hides global variable
* 1 for Unused local variable
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;
* 1 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: build-linux.sh [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
+sudo ln -s /opt/cmake/bin/cmake /usr/local/bin/cmake
+
+cd /vw
+mkdir build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE=Release -DWARNINGS=Off -DDO_NOT_BUILD_VW_C_WRAPPER=On
+NUM_PROCESSORS=$(cat nprocs.txt)
+make all -j ${NUM_PROCESSORS}
+make test_with_output
 cd ..
```
Comment (tyclintw) [inline build-linux.sh]: [nit] we should be using pushd and popd for these type of directory moves.

File: build-linux.sh [hunk 2]
```diff
@@ -1,17 +1,34 @@
...
 
-make all
-make python
-mvn clean test -f java/pom.xml
-make test
-make rl_clientlib_test
-cd test
-./test_race_condition.sh
+sudo apt remove --yes --force-yes cmake
+
```
Comment (tyclintw) [inline build-linux.sh]: We shouldn't be blindly removing cmake. Can we add a "cmake --version" check?

**Comment by JohnLangford** [comment]
This pull request **introduces 6 alerts** when merging 49d80b465d56aeca1cd233f4a1d5bbe305942392 into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-39c3a983c6ee5e8f49a70be5fc6b81560583665c)

**new alerts:**

* 2 for Local variable hides global variable
* 1 for Unused import
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: build-linux.sh (hunk 2 above)
Comment (jackgerrits) [reply build-linux.sh]: We control the image and so we know the cmake version is not current enough(3.5 is still very old) to be used. We would need to extract the version out of `cmake --version`  for that change to be useful and I don't think it is worth the work

**Comment by jackgerrits** [comment]
The 6 remaining alerts seem to be for files unchanged by this pr.

**Comment by JohnLangford** [comment]
@arielf @Scott-Graham-Bose can you take a look at the cmake patch?  It's a big change rewriting the build process.

**Comment by JohnLangford** [comment]
This seems to be parallelizing to hard on my laptop, causing swappage during build.

**Comment by JohnLangford** [comment]
This pull request **introduces 6 alerts** when merging 79b28878fc44ab292e7cdbb02152b03dd82ec928 into a18d3abde1c857720282c3e55325f0ef2fe3acc9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-9f0fc0f1798b3f8d44807605c6237781bd9b74bf)

**new alerts:**

* 2 for Local variable hides global variable
* 1 for Unused import
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
Issues:
1) Build doesn't seem to understand the number of processors.
2) Add targets for profiling, valgrind, static.
3) remove ffast-math from valgrind profile (fixing an inconsistency)
4) Fix static compile
5) Make it so install headers are automatically headers
6) Do we need disambiguators for else() in the vowpalwabbit directory?

File: java/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,55 @@
+find_package(JNI)
```
Comment (JohnLangford) [inline java/CMakeLists.txt]: @jmorra @deaktator can you look at the cmake build for java here?

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,53 @@
+add_subdirectory(unit_test)
```
Comment (JohnLangford) [inline test/CMakeLists.txt]: @arielf These are the change to invocation of tests.

**Comment by JohnLangford** [comment]
@eisber you may also be interested in this patch.

**Comment by eisber** [comment]
I already skipped through ☺

Thanks

**Comment by jackgerrits** [comment]
Disambiguators for else are not needed, they are used by CMake to warn if if-endifs do not match

**Comment by jackgerrits** [comment]
@JohnLangford I believe I have resolved the issues we came across today

**Comment by JohnLangford** [comment]
This pull request **introduces 6 alerts** when merging 8c1013eccf7202083b60a7227a1d01959f7a849d into 05cca528c58829225406131d927e2601e90d2b7c - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-2d9aaacfb47ac01a41f5be0cb649524a68e33042)

**new alerts:**

* 2 for Local variable hides global variable
* 1 for Unused import
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 6 alerts** when merging 8537b54678fa6ca1984b34f7a164b7ea12979503 into f11d32385ed8faa630d6c7cc5e705af89456b81a - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-b45d59723c5624b404813444a365d8f1e9d4f223)

**new alerts:**

* 2 for Local variable hides global variable
* 1 for Unused import
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 6 alerts** when merging 8537b54678fa6ca1984b34f7a164b7ea12979503 into ad60f867e30d3e03983440ca3bca74197e203824 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-007fac12941e3cabb03160195d092b20d83fb6cc)

**new alerts:**

* 2 for Local variable hides global variable
* 1 for Unused import
* 1 for No virtual destructor
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 3 alerts** when merging 7d1e2ffada32825ec5149c544e2c57820eee4025 into 1ea72b2c7f7206c3e7515a57cfb71468779d49f9 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-4ebce6480a514aeb77e2f08c9ad50c3697438f4e)

**new alerts:**

* 1 for Local variable hides global variable
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 3 alerts** when merging 163c58fbf25cb847416dce308aa98751189601df into acd910e4c802e9487b997a66472cd6cefb844360 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-64ddd464850cb3c1a50c960def69831f6298028d)

**new alerts:**

* 1 for Local variable hides global variable
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 3 alerts** when merging f9554f42888377315b053ce4f0fe4cb31796eb1d into 72f6f430f79b6eebb1cd87c056ece6313d66f23b - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-624a92dce41967dc0e1086dd2898d38014ca681c)

**new alerts:**

* 1 for Local variable hides global variable
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by JohnLangford** [comment]
This pull request **introduces 3 alerts** when merging 75ee3d0a9862996653d7b845a67b4cc5b3105a14 into fb8d4887d7feadca14228ce4de2fdb03256e6ee6 - [view on LGTM.com](https://lgtm.com/projects/g/JohnLangford/vowpal_wabbit/rev/pr-4f92a2e507ca71f64e700985ecd85eb8b86b47e2)

**new alerts:**

* 1 for Local variable hides global variable
* 1 for Comparison result is always the same
* 1 for Short global name

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**